Introduce `TermTree` and `LogicTree` classes in place of `is_expr`.

hameerabbasi opened the pull request:
Closes #23

willow-ahrens commented on src/finch/autoschedule/compiler.py:
```diff
@@ -35,19 +37,19 @@ def get_structure(
-                get_structure(lhs, fields, aliases), get_structure(arg, fields, aliases)
-            )
+            arg_2 = get_structure(arg, fields, aliases)
+            assert isinstance(arg_2, LogicTree)
```
It would be okay for this to be a LogicNode

hameerabbasi commented on src/finch/symbolic/term.py:
```diff
@@ -14,56 +17,38 @@
         `x == x.make_term(x.head(), *x.children())`
         """
 
-    def __hash__(self) -> int:
-        """Return the hash value of the term."""
-        if self._hashcache is None:
-            self._hashcache = hash(
-                (0x1CA5C2ADCA744860, self.head(), tuple(self.children()))
-            )
-        return self._hashcache
-
-    def __eq__(self, other) -> bool:
-        if not isinstance(other, Term):
-            return NotImplemented
-        return self.head() == other.head() and self.children() == other.children()
```
Hashing and equality is already implemented by the `@dataclass(frozen=True, eq=True)` decorator. Therefore I chose not to have redundant implementations.

willow-ahrens replied:
Can you verify that eq=true,frozen=true will cache the hash of the object? I don't think it will, we should probably leave this implementation alone

willow-ahrens replied:
Reading the docs on it makes me think they aren't caching (the fields themselves aren't known to be immutable). Maybe we should be setting a flag that lets us override hash.

hameerabbasi commented on src/finch/finch_logic/nodes.py:
```diff
@@ -479,9 +399,12 @@ def children(self):
     def make_term(cls, head, *args):
```
`@dataclass` defines a default `__init__`, and it's annoying to overwrite [for frustrating reasons](https://www.pythonmorsels.com/customizing-dataclass-initialization/).

Additionally, when one of the members is a tuple, one can't simply do `ClassName(normal_member, tuple_member)` but instead we have to do `ClassName(normal_member, *tuple_member)` (which can get confusing real fast, also bad for typing reasons).

mtsokol replied:
Right, then let's stick to default `__init__` and override `make_term` as it is now.

hameerabbasi commented:
Okay, I've tightened up the types and addressed all the comments above. PyTest passes, mypy fails for one line, which I suspect is incorrect but is never hit in the tests.

```
src/finch/autoschedule/compiler.py:82: error: Generator has incompatible item type "Immediate"; expected "Field"  [misc]
```

which points to

https://github.com/finch-tensor/finch-tensor-lite/blob/71bf16003401f32213c2b3d9e7367c892d55ff30/src/finch/autoschedule/compiler.py#L80-L83

But `Table` should always have a `tuple[Field, ...]` as its last argument, not `tuple[Immediate, ...]`. Even the docstring seems to imply that:

> ```
> Represents a logical AST expression for a tensor object `tns`, indexed by fields
>     `idxs...`. A table is a tensor with named dimensions.
> 
>     Attributes:
>         tns: The tensor object.
>         idxs: The fields indexing the tensor.
> ```

mtsokol reviewed (commented):
Last two comment from my side.

willow-ahrens reviewed (changes requested):
We're on the right track! Thanks for your help with this PR, it's improving the code substantially. The biggest request I have is to leave hash alone, set `eq=false`, and add eq and hash methods for the leaf nodes. I think this makes equality and hashing much more explicit and overrideable, they're important methods that we'll want to modify now and in the future.

willow-ahrens commented:
@hameerabbasi, I've edited the comments in my review a little bit regarding the hash. My current suggestion is to move our optimized `__hash__` and `__eq__` to TermTree, set `eq=false` for TermTree subclasses, and leave `eq=true` for TermTree and Leaf nodes like immediate, deferred, field, and alias. I think this will be a minimal change to keep a nice hashing optimization until we eventually get around to more advanced IR optimizations.

mtsokol commented on src/finch/autoschedule/optimize.py:
```diff
@@ -82,6 +87,24 @@ def rule_1(node):
     return Rewrite(PostWalk(Chain([rule_0, rule_1])))(root)
 
 
+@overload
+def _lift_subqueries_expr(  # type: ignore[overload-overlap]
+    node: Subquery, bindings: dict[LogicNode, LogicNode]
+) -> LogicExpression: ...
+
+
+@overload
+def _lift_subqueries_expr(
+    node: LogicTree, bindings: dict[LogicNode, LogicNode]
+) -> LogicTree: ...
+
+
+@overload
+def _lift_subqueries_expr(
+    node: LogicNode, bindings: dict[LogicNode, LogicNode]
+) -> LogicNode: ...
```
Let's leave it then.

mtsokol commented on src/finch/finch_logic/nodes.py:
```diff
@@ -366,29 +255,20 @@ class Reformat(LogicNode):
-    def get_fields(self):
+    def get_fields(self) -> list[Field]:
         """Returns fields of the node."""
+        assert isinstance(self.arg, LogicExpression)
```
Is it needed? `arg: LogicExpression` is already typed as `LogicExpression`.

mtsokol commented on src/finch/finch_logic/nodes.py:
```diff
@@ -254,25 +172,16 @@ class Aggregate(LogicNode):
-    def get_fields(self):
+    def get_fields(self) -> list[Field]:
         """Returns fields of the node."""
+        assert isinstance(self.arg, LogicExpression)
```
Same question as before.

mtsokol reviewed (commented):
From my side LGTM, but let's wait for @willow-ahrens approval.

willow-ahrens reviewed (approved):
LGTM! I'll make issues for make_term and hashing to happen separately.

willow-ahrens commented:
Are we good to merge?

mtsokol commented:
It's all green now - merging!
